Update setup instructions

lmoureaux opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

User setup instructions are currently broken. This PR is about rewriting them.

I used Brian's script as a base, and modified it to run on every tested machine. I also added a few checks. The new script is [here](https://github.com/cms-gem-daq-project/sw_utils/pull/1).

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context

From most to least relevant: 

* https://github.com/cms-gem-daq-project/sw_utils/pull/1
* Issue #96
* Jared's Gists:
    * https://gist.github.com/jsturdy/a82a6b7b7d497ae8ef98ee736119c6b5
    * https://gist.github.com/jsturdy/4f9bccbdd18d953f73777e492a71fd24
* Brian's Gists:
    * https://gist.github.com/bdorney/5f3965bb8b26858a91ef8087137d1ec9
    * https://gist.github.com/bdorney/e91ade728f7e6f37d1820299fba1d9c5
* Testing requested for PR #105
* Issue #107

## How Has This Been Tested?

The instructions have been tested in the following environments:

* `lxplus` (`slc6`)
* `gem904daq01` (`cc7`)
* `gem904daq03` (`slc6`)
* `gem-dqm01` (`slc6`)

Before starting, I removed all `pip` packages installed in `.local`. I used my favorite shell, `bash`.

The testing itself consisted of:

 * Setting up the env
 * Checking that `anaUltraScurve.py -h` ran correctly. This test isn't as dumb as it may seem at first: `anaUltraScurve.py` imports from `ROOT`, `numpy`, `gempython.utils` and `gempython.gemplotting` before even parsing its arguments
 * Logging out and in again to clear any environment variable
 * Re-activating the environment
 * Checking `anaUltraScurve.py -h` again

### Planned work

* Update developer docs (once user docs are stable)

### Help needed *[obsolete]*

I briefly tested the setup script with `zsh`, and it's broken (doesn't call `pip` and `virtualenv`). I'm quite sure other shells are affected. I'd like to get:

* A list of shells that should be supported
* Help to make the script more generic -- I'm only familiar with `bash`

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

bdorney commented on README.md:
````diff
@@ -59,33 +59,43 @@ Table of Contents
 ```
-source $BUILD_HOME/cmsgemos/setup/paths.sh
-source $BUILD_HOME/gem-plotting-tools/setup/paths.sh
+source gem-plotting-tools/setup.sh -c 0.3.1 -g 1.0.0 -G 5 -v 2.0.0 -V 3
````
I am not sure this script should not belong to this repository.  Ideally it should belong to a future, not yet created, meta repo which will handle the job of checking cross-repo compatibility at both SW & FW level (e.g. Repo A version X.Y.Z is compatible with repo B version A.B.C and expects FW version 1.2.3).  I am not sure it make sense to have a tool the user will use to get the code inside the code; especially one that takes the other repos.

Until this repo is created by @jsturdy and @mexanick I think we need another strategy.  My view would be to create the above mentioned repo and have this script be the first commit.  Then it could be expanded upon from there.  The instructions would then still reference this script but *not* in the repo.

Perhaps @jsturdy or @mexanick could comment on this issue.

jsturdy replied:
It should not, especially as the only dependency from `gemplotting` is on the `utils` of `cmsgemos_gempython` (i.e., *no* dependence on `vfatqc`)

Furthermore, for a pure `USER`, no checking out of a `git` repository should ever need to be done.

To this end, I'm also prototyping setting up a `PyPi` repository that will have the packages pushed to, either on the official site, or setting up a private one as part of the CERN hosted `PaaS` ecosystem being developed for our uses

bdorney replied:
Okay for the short-term I would suggest to follow @jsturdy's statement.  To that end for the short-term:

1. remove `setup.sh` from the repo,
2. restore the previous `*.sh` scripts,
3. update the `README.md` to download the script at the gist and have the user follow those,
4. do not close related setup/documentation issues as we will return to this later.

This will allow the continued analysis of the channel loss study on p5 data and will keep open the further developments.

bdorney commented on README.md:
````diff
@@ -542,7 +552,8 @@ The format of this input file should follow the [Two Column Format](two-column-f
 ```
-gemSCurveAnaToolkit.py -ilistOfScanDates_Scurve.txt -v0 -s29 --anaType=scurveAna -c --summary --drawLeg 
+gemSCurveAnaToolkit.py -ilistOfScanDates_Scurve.txt -v0 -s29 --anaType=scurveAna -c --summary --drawLeg
+
````
un-necessary white space, eliminate.

bdorney commented on setup.sh:
```diff
@@ -0,0 +1,289 @@
+            return 1;;
+    esac
+done
+unset OPTIND
```
why `unset` this here? It's not being exported and should only be valid inside this execution.

lmoureaux replied:
Not when the script is being sourced, which is its primary usage

bdorney commented on setup.sh:
```diff
@@ -0,0 +1,289 @@
+SYSTEM_INFO="$(uname -a)"
+KERNEL_VERSION="$(uname -r)"
+VIRTUALENV=virtualenv
+PIP=pip
```
What is the rationale behind this assignment?  Why is it necessary?

bdorney commented on setup.sh:
```diff
@@ -0,0 +1,289 @@
+helpstring="Usage: source $BASH_SOURCE [options]
```
Why have you removed the shebang on this script?

lmoureaux replied:
The script is supposed to be sourced, in which case the shebang has no effect. I added checks to prevent the user from running it (because of [this comment](https://github.com/cms-gem-daq-project/gem-plotting-tools/issues/96#issuecomment-392785774)).

bdorney commented on setup/createVirtualEnvPy26.sh:
```diff
@@ -1,29 +0,0 @@
-#!/bin/zsh
```
while this script may not be used presently I am not sure it should be removed.  Generally reducing functionality is unwise.

bdorney commented on setup/paths.sh:
```diff
@@ -1,81 +0,0 @@
-# Setup cmsgemos
```
same

lmoureaux commented:
>> I briefly tested the setup script with zsh, and it's broken (doesn't call pip and virtualenv). I'm quite sure other shells are affected. I'd like to get:
>
> Which script are you referring to, and what is the calling syntax and terminal output? The script shown at my gist above works without issue in zsh as that's my default $SHELL.

My modified script, `setup.sh`. The following lines may need to be updated:

* The two lines that use `$BASH_SOURCE`, though they're not critical
* Lines calling `pip` and `virtualenv` using `$VIRTUALENV` and `$PIP` -- these are critical

lmoureaux commented:
## Point 5 update

Here are preliminary instructions. `anaUltraScurve.py -h` runs fine, but I haven't tested my updated script at `lxplus` and 904 (yet).

I keep the comments above in mind, just processing one issue at a time :snail:

@jsturdy Is installing `PySocks` system-wide an option?

### Setup at Point 5

Due to the limited Internet access, the setup at Point 5 is more involved.

#### To be done once

Download a package to enable the use of SOCKS proxies in Python:

```
ssh cmsusr wget https://files.pythonhosted.org/packages/53/12/6bf1d764f128636cef7408e8156b7235b150ea31650d0260969215bb8e7d/PySocks-1.6.8.tar.gz
```

Install it:

```
pip install --user PySocks-1.6.8.tar.gz
```

#### To be done every time you create a new environment

Download the `cmsgemos` package manually:

```
scp lxplus.cern.ch:/afs/cern.ch/user/s/sturdy/public/cmsgemos_gempython-0.3.1.tar.gz .
```

Create a SOCKS proxy that will allow `pip` to reach the outer world:

```
PORT=5000
ssh -D *:$PORT lxplus.cern.ch -N -f
```

If you get an error saying `bind: Address already in use`, try with `PORT=5001`, `5002`, ...

> **Note**
> The proxy expires after some time. Just create it again if `pip` complains about the network being unreachable.

Define `$ELOG_PATH`:

```
export ELOG_PATH=/your/favorite/elog/path
```

Then execute:

```
source gem-plotting-tools/setup.sh -c 0.3.1 -g 1.0.0 -G 5 -v 2.0.0 -V 3 -P $PORT
```

Congratulations, you are done! You can use the usual commands to `deactivate` your `virtualenv` and activate it again.

bdorney commented:
This part:

>To be done once
Download a package to enable the use of SOCKS proxies in Python:
ssh cmsusr wget https://files.pythonhosted.org/packages/53/12/6bf1d764f128636cef7408e8156b7235b150ea31650d0260969215bb8e7d/PySocks-1.6.8.tar.gz
Install it:
pip install --user PySocks-1.6.8.tar.gz
To be done every time you create a new environment

Could be scripted via something like:

```
PYSOCKS=$(pip list --user | grep "PySocks")
if [ ! -n "$PYSOCKS" ]
then

    # run the commands to make a user install of PySocks
    
fi
```

Of course a system wide installation could be done but this again depends on the sysadmin and if the package could be added at time of machine setup (e.g. through puppet)

This part:

>Create a SOCKS proxy that will allow pip to reach the outer world:
PORT=5000
ssh -D *:$PORT lxplus.cern.ch -N -f
If you get an error saying bind: Address already in use, try with PORT=5001, 5002, ...
Note
The proxy expires after some time. Just create it again if pip complains about the network being unreachable.

Could also be scripted.  The creation of the proxy should return some exit code which can be checked in a loop.  For example perform this in a `do...while(exit code not zero)` and each iteration of the loop increment the port number until a successful port is found.  Although maybe someone has a good argument on why automatic port assignment here is a bad idea (I could think of a few but perhaps in the 5000 range it's okay...).

lmoureaux commented:
Force-pushed updates:

* Rewrote history not to remove the old scripts
* Addressed/answered all but [one](https://github.com/cms-gem-daq-project/gem-plotting-tools/pull/109#issuecomment-394765331) of Brian's comments
* [The script](https://gist.github.com/lmoureaux/97c663d2ff63b977c5823a8e010c4e1d) works everywhere (updated the first post with the current state of testing)

Caveats:

* The problems with `zsh` have got worse
* Didn't touch my `.ssh/config` at P5 (it has some config to be able to reach `lxplus`, dunno how standard it is)
* Should maybe instruct the user to enter their `cmsusr` and `lxplus` passwords when asked

Next:

* Developer docs

jsturdy commented:
A workaround for differences between pure `bash` (`$BASH_SOURCE`) and `zsh` can be seen [here](https://github.com/cms-gem-daq-project/ctp7_modules/blob/1bca5c8930ccefa4b67a3a435654d38ba9a1859e/setup.sh)

lmoureaux commented:
> A workaround for differences between pure bash ($BASH_SOURCE) and zsh can be seen here

Thanks @jsturdy! The other problem I have with `zsh` is:

```
CMD="echo A"
$CMD
```
Output:
`bash`: `A`
`zsh: command not found: echo A`

bdorney commented:
> The problems with zsh have got worse

Could you describe?

>Thanks @jsturdy! The other problem I have with zsh is:
CMD="echo A"
$CMD
Output:
bash: A
zsh: command not found: echo A

In `zsh` the correct way to do this is via the use of arrays, see for example:

```
% CMD=(echo A)
% $CMD
A
```

For more details see the FAQ, specifically [section 3.1](http://zsh.sourceforge.net/FAQ/zshfaq03.html).

bdorney commented:
Additionally can you push the changes you made to `setup_gemdaq.sh` here:

https://github.com/cms-gem-daq-project/sw_utils

And then name it `setup_gemdaq.sh`.

I don't want `USER` to checkout a github repo but I think `setup_gemdaq.sh` is now sufficiently advanced where I would like to have some subversioning and history.  I'll fork it and add my own changes on top of that.

lmoureaux commented:
> Additionally can you push the changes you made to setup_gemdaq.sh here:

I'll `wget` the script directly from the repo. Should I use `master` or another branch?

bdorney commented:
> I'll wget the script directly from the repo. Should I use master or another branch?

Push it to `develop` and we will go from there.

lmoureaux commented:
>> I'll wget the script directly from the repo. Should I use master or another branch?
>
> Push it to develop and we will go from there.

I meant in the README instructions.

Additionally I'm not allowed to push to the new repo.

bdorney commented:
> I'll wget the script directly from the repo

The script is already there, my interest is to incorporate the changes you made, and also some changes I've made based on some feedback from @jsturdy

lmoureaux commented:
* Updated the script for `zsh` support (tested on `lxplus`).
* Updated the instructions to point to the new `sw_utils` repo (not valid until https://github.com/cms-gem-daq-project/sw_utils/pull/1 reaches `master`)

lmoureaux commented:
Removing the WIP tag as testing is now complete. Dev docs update will be in another PR.

bdorney commented on README.md:
```diff
@@ -59,33 +60,89 @@ Table of Contents
+
+* `cmsgemos` version 0.3.1 (`-c 0.3.1`)
+* `gemplotting` version 1.0.0-dev5 (`-g 1.0.0 -G 5`)
+* `vfatqc` version 2.0.0-dev3 (`-v 2.0.0 -V 3`)
```
Change too:

---
* [cmsgemos](https://github.com/cms-gem-daq-project/cmsgemos/tags) version X.Y.Z (`-c X.Y.Z`)
* [gemplotting](https://github.com/cms-gem-daq-project/gem-plotting-tools/tags) version X.Y.Z-dev**A** (`-g X.Y.Z -G A`)
* [vfatqc](https://github.com/cms-gem-daq-project/vfatqc-python-scripts/tags) version X.Y.Z-dev**A** (`-v X.Y.Z -V A`)

Note that if a dev version is not to be used (normal) drop the -G option.

---

Again hard coding a given release should be avoided (in anticipation of future changes).

lmoureaux replied:
How do I connect it to the line above:

```
source setup_gemdaq.sh -c 0.3.1 -g 1.0.0 -G 5 -v 2.0.0 -V 3
```

Should I use `A.B.C`, `D.E.F-devG`, `H.I.J-devK` ? Or worse, `X1.X2.X3`, `Y1.Y2.Y3-devY4`, `Z1.Z2.Z3-devZ4` ? I seems to me that both are less readable than a concrete example.

bdorney replied:
I am not asking you to remove the concrete example but to provide additional information in the form of a bullet point list with links to the relevant tag areas on GitHub with syntax shown above.

lmoureaux replied:
If you meant "add" and not "change", where should I add it? How do I connect it to:

```
source setup.sh -c <cmsgemos tag> -g <gem-plotting tag> -G <gem-plotting dev version optional>
```

Sorry but I really don't understand how you want me to fit this new bullet list into the text.

bdorney commented on README.md:
````diff
@@ -59,33 +60,89 @@ Table of Contents
+To re-enable the python env, source the script again:
+
+```
+source setup.sh
````
I think the name is now `setup_gemdaq.sh`

bdorney commented on README.md:
````diff
@@ -59,33 +60,89 @@ Table of Contents
 ```
 
-If you are working on a 904 machine, regardless if it is the first time you are executing the above commands or not, the default python `virtualenv` available on the 904 NAS for your operating system (e.g. `cc7` or `slc6`) will be enabled.
+Remove and download the setup script:
````
why remove...?  If this is to make sure they have the most up-to-date version add this statement here.  e.g. 

> Remove and download the setup script to ensure you have the most up-to-date version.

bdorney commented on README.md:
````diff
@@ -59,33 +60,89 @@ Table of Contents
+Then execute:
+
+```
+source setup.sh -c 0.3.1 -g 1.0.0 -G 5 -v 2.0.0 -V 3 -P $PORT
````
Same comment as [here](https://github.com/cms-gem-daq-project/gem-plotting-tools/pull/109/files#r193453084)

bdorney commented on README.md:
````diff
@@ -542,7 +599,7 @@ The format of this input file should follow the [Two Column Format](two-column-f
 
 ```
-gemSCurveAnaToolkit.py -ilistOfScanDates_Scurve.txt -v0 -s29 --anaType=scurveAna -c --summary --drawLeg 
+gemSCurveAnaToolkit.py -ilistOfScanDates_Scurve.txt -v0 -s29 --anaType=scurveAna -c --summary --drawLeg
````
I cannot see what changed here...?

lmoureaux replied:
Turned `U+2028 LINE SEPARATOR` into an ASCII newline (not that I was aware of it... but remember I had an issue with a spurious newline here).

Since it's the only place where this character was present, I would say this is a fix. I can reintroduce it, but not every tool may work well with such an exotic character.

*\<troll>If you ask me to put it back, next time I may introduce other funny characters such as ͘ı orａ\</troll>*

bdorney replied:
No I was just trying to understand.  Evidently GitHub is just not rendering the character in question.  Proceed with other changes as discussed above.  Thanks.

bdorney commented on README.md:
```diff
@@ -59,33 +60,106 @@ Table of Contents
+>
+> * `cmsgemos` version 0.3.1 (`-c 0.3.1`)
+> * `gemplotting` version 1.0.0-dev5 (`-g 1.0.0 -G 5`)
+> * `vfatqc` version 2.0.0-dev3 (`-v 2.0.0 -V 3`)
```
Remove leading `>` as this will be rendered by MarkDown as quote mode (not applicable here) rather than preformatted mode (desired).

bdorney replied:
Yes, I checked it on your GitHub page and removed the comment as can be seen.

bdorney commented on README.md:
```diff
@@ -59,33 +60,106 @@ Table of Contents
+>
+> * `cmsgemos` version 0.3.1 (`-c 0.3.1`)
+> * `gemplotting` version 1.0.0-dev5 (`-g 1.0.0 -G 5`)
+> * `vfatqc` version 2.0.0-dev3 (`-v 2.0.0 -V 3`)
```
same

bdorney commented on README.md:
````diff
@@ -59,33 +60,106 @@ Table of Contents
 ```
 
-If this is the first time you are executing the above commands on `lxplus` it will create a `python v2.7` `virtualenv` for you.  It may take some time to download the necessary packages so be patient and do not interrupt the installation.  To disable this python env execute:
+If a development version is not to be used (normal case), you can drop the options with capital letters (`-G`, `-V`). If this is the first time you are executing the above command, it will create a Python `virtualenv` for you and install the `cmsgemos`, `gemplotting` and `vfatqc` packages. It may take some time to download them, so be patient and do not interrupt the installation.
````
Please take a look at https://github.com/cms-gem-daq-project/gem-plotting-tools/pull/109#discussion_r193454753.

I would like specifically this type of system inserted so the user has:

1. The location of where to see the tags, and
2. An explicit understanding of the tag and dev versioning and how that relates.

lmoureaux replied:
1. Oops, didn't notice the links
2. Can you be more specific? Do you mean an understanding of the whole versionning system?

bdorney replied:
It’s more that you have not shown what is requested. Please copy/paste exactly what I wrote, which I think will give a descriptive generalization, in the now “outdated” comment in the location where it was originally referenced.  Thanks!

bdorney commented on README.md:
```diff
@@ -91,9 +91,9 @@ If a development version is not to be used (normal case), you can drop the optio
-> * `vfatqc` version 2.0.0-dev3 (`-v 2.0.0 -V 3`)
+> * [cmsgemos](https://github.com/cms-gem-daq-project/cmsgemos/tags) version 0.3.1 (`-c 0.3.1`)
+> * [gemplotting](https://github.com/cms-gem-daq-project/gem-plotting-tools/tags) version 1.0.0-dev5 (`-g 1.0.0 -G 5`)
+> * [vfatqc](https://github.com/cms-gem-daq-project/vfatqc-python-scripts/tags) version 2.0.0-dev3 (`-v 2.0.0 -V 3`)
```
Adding the links as you’ve done so here don’t really make sense since you’re providing the general link but a specific tag. Please see my original request for clarification.

bdorney commented on README.md:
```diff
@@ -155,9 +155,9 @@ If a development version is not to be used (normal case), you can drop the optio
-> * `vfatqc` version 2.0.0-dev3 (`-v 2.0.0 -V 3`)
+> * [cmsgemos](https://github.com/cms-gem-daq-project/cmsgemos/tags) version 0.3.1 (`-c 0.3.1`)
+> * [gemplotting](https://github.com/cms-gem-daq-project/gem-plotting-tools/tags) version 1.0.0-dev5 (`-g 1.0.0 -G 5`)
+> * [vfatqc](https://github.com/cms-gem-daq-project/vfatqc-python-scripts/tags) version 2.0.0-dev3 (`-v 2.0.0 -V 3`)
```
Same

bdorney commented:
Also upon further reflection we should probably remove all references to `vfatqc` here since there is *no* dependence of `gem-plotting` on `vfatqc`.  e.g. if a `USER` wants to do only analysis they don't need `vfatqc` at all.

lmoureaux commented:
@bdorney, I'm quite sure you'll have an opinion about how the `-pre` releases should be handled. Currently we only give instructions for stable and `-dev` releases.

bdorney commented:
> @bdorney, I'm quite sure you'll have an opinion about how the -pre releases should be handled. Currently we only give instructions for stable and -dev releases.

What's the use case?  Naively I guess a `USER` is using a stable release and a `developer` is using either a stable or a `dev` release.  Try to ask @jsturdy what a `pre` release would be doing and who should be using it.  I assume only a `sysadmin` would be using a `pre` release and xcheck'ing it installs correctly but I could be mistaken.

lmoureaux commented:
> `travis-ci` build did not complete
Travis checks aren't reliable, they fail every now and then because they fail to fetch the Docker images. IIRC there's a "Retry" button somewhere (or is it on Gitlab?), pushing it might "fix" the issue.
